### PR TITLE
Add liquidity processor and producer for pdex state only

### PR DESCRIPTION
**Pull request by tin-incognito**
*Change*

1. Instruction format
2. Instruction handler

*Add*

1. Verify instruction for processor
2. Add liquidity update PDEX state for processor and producer
3. Update remainTxs format

### CODE REVIEW

File: blockchain/pdex/pool_pair_state.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+func (p *PoolPairState) addShare(key string, amount, beaconHeight uint64) string {
+	nfctID := genNFT(key, p.state.CurrentContributionID(), beaconHeight)
+	share := statedb.NewPdexv3ShareStateWithValue(amount)
+	p.shares[nfctID] = *share
+	p.state.SetCurrentContributionID(p.state.CurrentContributionID() + 1)
+	return nfctID
+}
+
+func genNFT(key string, id, beaconHeight uint64) string {
```
Comment (dat-incognito-org) [inline blockchain/pdex/pool_pair_state.go:135]: id & beaconHeight are not of fixed lengths, this has the risk of pre-image collision

File: dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	"github.com/incognitochain/incognito-chain/common"
+)
+
+type Pdexv3PoolPairState struct {
+	token0ID              string
```
Comment (dat-incognito-org) [inline dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go:12]: using the type `common.Hash` for token IDs instead makes them well-defined. What are the benefits of using string here ?

File: relaying/btc/btc-blocks-mainnet/metadata/CURRENT [hunk 3]
```diff
@@ -0,0 +1 @@
+MANIFEST-000000
```
Comment (dat-incognito-org) [inline relaying/btc/btc-blocks-mainnet/metadata/CURRENT:1]: are these database files ?

File: blockchain/pdex/interface.go [hunk 4]
```diff
@@ -20,8 +23,8 @@ type StateReader interface {
 	WaitingContributionsV1() map[string]*rawdbv2.PDEContribution
 	DeletedWaitingContributionsV1() map[string]*rawdbv2.PDEContribution
 	PoolPairsV1() map[string]*rawdbv2.PDEPoolForPair
-	WaitingContributionsV2() map[string]Contribution
-	DeletedWaitingContributionsV2() map[string]Contribution
+	WaitingContributionsV2() map[string]statedb.Pdexv3ContributionState
+	DeletedWaitingContributionsV2() map[string]statedb.Pdexv3ContributionState
```
Comment (cuong-incognito) [inline blockchain/pdex/interface.go:27]: So after all refactors, we still have to use the different functions for state versions 😢

File: relaying/btc/btc-blocks-mainnet/metadata/CURRENT (hunk 3 above)
Comment (tin-incognito) [reply relaying/btc/btc-blocks-mainnet/metadata/CURRENT:1]: Hmm ok I will check this maybe something not right here

File: dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go (hunk 2 above)
Comment (tin-incognito) [reply dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go:12]: This struct will store directly to statedb by byte so I want to use basic data type of Golang

File: blockchain/pdex/pool_pair_state.go (hunk 1 above)
Comment (tin-incognito) [reply blockchain/pdex/pool_pair_state.go:135]: Hmmm so we need to fix length of id and beaconHeight like number 1 to 9?

File: blockchain/pdex/interface.go (hunk 4 above)
Comment (tin-incognito) [reply blockchain/pdex/interface.go:27]: Yes this is maybe my problem. I still can't find the way to get state with different data structure return. But hey, this is only for getter function which only will be used by RPC and it need to know which version is being called

File: blockchain/pdex/pool_pair_state.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+		make(map[string]statedb.Pdexv3ShareState),
+		make(map[string]map[string]uint64),
+	)
+}
+
+func NewPoolPairState() *PoolPairState {
+	return &PoolPairState{
+		shares:      make(map[string]statedb.Pdexv3ShareState),
+		tradingFees: make(map[string]map[string]uint64),
+	}
```
Comment (cuong-incognito) [inline blockchain/pdex/pool_pair_state.go:47]: can't find the initialized value of `state`

File: blockchain/pdex/staking_pool_state.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package pdex
+
+type StakingPoolState struct {
+	liquidity        uint64
+	stakers          map[string]uint64 // nfst -> amount staking
+	currentStakingID uint64
```
Comment (cuong-incognito) [inline blockchain/pdex/staking_pool_state.go:6]: so after the trading fee is collected from the block, do we have to manually distribute the reward for every stakers?
I didn't see any property for reward storage. Besides, I don't think it is an efficient way.

File: blockchain/pdex/utils.go [hunk 7]
```diff
@@ -580,3 +581,37 @@ func InitStateFromDB(
...
+func generatePoolPairKey(token0Name, token1Name, txReqID string) string {
+	if token0Name <= token1Name {
+		return strings.Join([]string{token0Name, token1Name, txReqID}, "-")
+	}
+	return strings.Join([]string{token1Name, token0Name, txReqID}, "-")
+}
+
+//amplifier >= 10000
+func calculateVirtualAmount(amount0, amount1 uint64, amplifier uint) (uint64, uint64) {
+	if amplifier == metadataPdexV3.DefaultAmplifier {
```
Comment (cuong-incognito) [inline blockchain/pdex/utils.go:594]: Should choose a more meaningful name for the base amplifier. Ex: `BaseAmplifier`, `DenominatingAmplifier`.
Besides, I think we should use the `big.Int` type for the virtual amount (take an example, token DAI in our chain has 9 decimal digits, so if we have 100M DAI in liquidity = 10^17 token units, in a stable coin pair, the amplifier rate is about 1000 => we need 10^20 token units to handle DAI virtual amounts. It exceeds the range value of `uint64`

File: rpcserver/http_pdexv3.go [hunk 8]
```diff
@@ -282,7 +282,6 @@ func (httpServer *HttpServer) createRawTxAddLiquidityV3(
 	if len(keyWallet.KeySet.PrivateKey) == 0 {
 		return nil, isPRV, rpcservice.NewRPCError(rpcservice.RPCInvalidParamsError, fmt.Errorf("private key length not valid: %v", keyWallet.KeySet.PrivateKey))
```
Comment (cuong-incognito) [inline rpcserver/http_pdexv3.go:283]: Please bear in mind that we should never log the private key of the user in any case.

File: blockchain/pdex/pool_pair_state.go (hunk 1 above)
Comment (cuong-incognito) [reply blockchain/pdex/pool_pair_state.go:135]: how about joining string with commas as splits

Comment (cuong-incognito) [reply blockchain/pdex/pool_pair_state.go:135]: how about joining string with commas as splits

File: rpcserver/http_pdexv3.go (hunk 8 above)
Comment (tin-incognito) [reply rpcserver/http_pdexv3.go:283]: ok

File: blockchain/pdex/interface.go (hunk 4 above)
Comment (tin-incognito) [reply blockchain/pdex/interface.go:27]: I got solutions for this problems

File: blockchain/pdex/staking_pool_state.go (hunk 6 above)
Comment (tin-incognito) [reply blockchain/pdex/staking_pool_state.go:6]: any suggestions?

File: blockchain/pdex/utils.go (hunk 7 above)
Comment (tin-incognito) [reply blockchain/pdex/utils.go:594]: ok

File: blockchain/pdex/pool_pair_state.go (hunk 1 above)
Comment (dat-incognito-org) [reply blockchain/pdex/pool_pair_state.go:135]: just call `common.Uint64ToBytes` on each, outputs will be fixed length

File: dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go (hunk 2 above)
Comment (dat-incognito-org) [reply dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go:12]: highly doubt. The type `common.Hash` is an alias of `[32]byte`, which is just as primitive as string

File: blockchain/pdex/pool_pair_state.go (hunk 1 above)
Comment (dat-incognito-org) [reply blockchain/pdex/pool_pair_state.go:135]: - In the case the desired output here is a readable, unique name, simply using a `-` delimiter is acceptable
- I believe `key` has fixed length. Please add a comment specifying length of `key` for reviewers later on

File: dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go (hunk 2 above)
Comment (tin-incognito) [reply dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go:12]: Ok that's make sense for hash

File: blockchain/pdex/staking_pool_state.go (hunk 6 above)
Comment (cuong-incognito) [reply blockchain/pdex/staking_pool_state.go:6]: I think we should take the approach of distributing staking rewards the same as trading fee distributing for LP.
Take this as an example:
```
stakers map[string]StakingInfo // nfst -> staking info
  
type StakingInfo struct {  
   stakingAmount  uint64
   uncollectedReward uint64
   lastUpdatedBeaconHeight uint64
}
```

Comment (tin-incognito) [reply blockchain/pdex/staking_pool_state.go:6]: sounds good

Comment (tin-incognito) [reply blockchain/pdex/staking_pool_state.go:6]: Please also check trading fee fields of poolpairstate, if there is something need to be change with this structure

File: dataaccessobject/statedb/stateobject_pdexv3_share.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	"github.com/incognitochain/incognito-chain/common"
+)
+
+type Pdexv3ShareState struct {
+	amount uint64
```
Comment (cuong-incognito) [inline dataaccessobject/statedb/stateobject_pdexv3_share.go:12]: About `Pdexv3ShareState` can you move `trading fee` from `PoolPairState` to here, we also need `lastUpdatedBeaconHeight`.
```
type Pdexv3ShareState struct {
	amount uint64
	tradingFees map[string]uint64
	lastUpdatedBeaconHeight uint64
}
```

File: dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	"github.com/incognitochain/incognito-chain/common"
+)
+
+type Pdexv3PoolPairState struct {
```
Comment (cuong-incognito) [inline dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go:11]: PoolState still lacks fields that need to track trading fees. I think I will update this state after your PR.

Comment (tin-incognito) [reply dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go:11]: I'm writing it please check blockchain/pdex package for more detail

Comment (tin-incognito) [reply dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go:11]: Wair for update this PR and review it also with other comments

File: blockchain/pdex/pool_pair_state.go (hunk 1 above)
Comment (tin-incognito) [reply blockchain/pdex/pool_pair_state.go:135]: Done

File: dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go (hunk 2 above)
Comment (tin-incognito) [reply dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go:12]: Done

File: blockchain/pdex/staking_pool_state.go (hunk 6 above)
Comment (tin-incognito) [reply blockchain/pdex/staking_pool_state.go:6]: Done

File: rpcserver/http_pdexv3.go (hunk 8 above)
Comment (tin-incognito) [reply rpcserver/http_pdexv3.go:283]: Done

File: dataaccessobject/statedb/stateobject_pdexv3_share.go (hunk 9 above)
Comment (tin-incognito) [reply dataaccessobject/statedb/stateobject_pdexv3_share.go:12]: Done

File: dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go (hunk 10 above)
Comment (tin-incognito) [reply dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go:11]: Done

File: blockchain/pdex/pool_pair_state.go [hunk 11]
```diff
@@ -118,36 +115,51 @@ func (p *PoolPairState) updateReserveAndShares(
 	}
 	p.state.SetToken0RealAmount(p.state.Token0RealAmount() + amount0)
 	p.state.SetToken1RealAmount(p.state.Token1RealAmount() + amount1)
-	p.state.SetToken0VirtualAmount(p.state.Token0VirtualAmount() + amount0)
-	p.state.SetToken1VirtualAmount(p.state.Token1VirtualAmount() + amount1)
+
+	oldToken0VirtualAmount := p.state.Token0VirtualAmount()
```
Comment (dat-incognito-org) [inline blockchain/pdex/pool_pair_state.go:119]: do not cast uint64 to int64, it has the risk of overflow
`big.NewInt(0).SetUint64(amount0)` should be used instead
same for amount1

File: blockchain/pdex/utils.go [hunk 12]
```diff
@@ -590,9 +590,9 @@ func generatePoolPairKey(token0Name, token1Name, txReqID string) string {
 }
 
 //amplifier >= 10000
-func calculateVirtualAmount(amount0, amount1 uint64, amplifier uint) (uint64, uint64) {
+func calculateVirtualAmount(amount0, amount1 uint64, amplifier uint) (*big.Int, *big.Int) {
 	if amplifier == metadataPdexV3.DefaultAmplifier {
-		return amount0, amount1
+		return big.NewInt(int64(amount0)), big.NewInt(int64(amount1))
```
Comment (dat-incognito-org) [inline blockchain/pdex/utils.go:595]: do not cast uint64 to int64, use `.SetUint64()` instead

File: blockchain/pdex/pool_pair_state.go (hunk 11 above)
Comment (tin-incognito) [reply blockchain/pdex/pool_pair_state.go:119]: my bad. Thanks

File: rpcserver/http_pdexv3.go (hunk 8 above)
Comment (cuong-incognito) [reply rpcserver/http_pdexv3.go]: Please also fĩx logging private key at line 280

File: dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go (hunk 10 above)
Comment (cuong-incognito) [reply dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go:12]: I think we still lack the fields that I mentioned in the document.
```
lpFeesPerShare          map[string]uint64            // tokenID -> accumulated LP fee / total share
protocolFees            map[string]uint64            // tokenID -> accumulated uncollected protocol fee
stakingPoolFees         map[string]uint64.           // tokenID -> distributed fee for staking pools of the current block
```

Comment (tin-incognito) [reply dataaccessobject/statedb/stateobject_pdexv3_pool_pair.go:12]: Fee managerment will be update in other PR. This PR take so much time for review and block other PRs

File: relaying/btc/btc-blocks-mainnet/metadata/CURRENT (hunk 3 above)
Comment (tin-incognito) [reply relaying/btc/btc-blocks-mainnet/metadata/CURRENT:1]: Done

File: blockchain/pdex/pool_pair_state.go (hunk 5 above)
Comment (tin-incognito) [reply blockchain/pdex/pool_pair_state.go]: Done

File: blockchain/pdex/utils.go (hunk 7 above)
Comment (tin-incognito) [reply blockchain/pdex/utils.go:594]: Done

File: rpcserver/http_pdexv3.go (hunk 8 above)
Comment (tin-incognito) [reply rpcserver/http_pdexv3.go]: Done

File: blockchain/pdex/pool_pair_state.go (hunk 11 above)
Comment (tin-incognito) [reply blockchain/pdex/pool_pair_state.go:119]: Done

File: blockchain/pdex/utils.go (hunk 12 above)
Comment (tin-incognito) [reply blockchain/pdex/utils.go:595]: Done

**Review by dat-incognito-org — APPROVED** [review]

**Review by cuong-incognito — APPROVED** [review]

**Pull request merged**